### PR TITLE
Define new hooks for tunnel plugins

### PR DESCRIPTION
**Pull request by isaacroldan**
### WHY are these changes introduced?
In the process of migrating the ngrok tunnel to use hooks, we first need to define those hooks in cli-kit so that they can be the imported in any tunnel plugin.

Contributes to https://github.com/Shopify/shopify-cli-planning/issues/305

### WHAT is this pull request doing?
Add two new typed hooks:
- `tunnel_provider` for discoverability. Any plugin implementing this hooks means it can handle tunnels.
- `tunnel_start_${provider}` for execution. After discovering the plugins, each one will report what hook to use to start a tunnel, this hook is unique per plugin.

This approach allow us to have multiple tunnel plugins installed and decide at runtime which one we want to use for tunnelling depending on user input/selection without having incompatibilities.

### How to test your changes?
Nothing to test yet

### CODE REVIEW

**Review by pepicrft — APPROVED** [review]

File: packages/cli-kit/src/plugins.ts [hunk 1]
```diff
@@ -70,4 +88,4 @@ export type FanoutHookFunction<
 > = (
   this: Interfaces.Hook.Context,
   options: TPluginMap[TEvent]['options'] & {config: Interfaces.Config},
-) => Promise<Partial<PluginReturnsForHook<TEvent, TPluginName, TPluginMap>>>
+) => Promise<PluginReturnsForHook<TEvent, TPluginName, TPluginMap>>
```
[the remark this answers is not in the document]
Comment (shauns) [reply packages/cli-kit/src/plugins.ts:91]: I think that change is fine 👍  -- the partials in this area are mainly meant to be about not relying on a plug-in being installed.

File: packages/cli-kit/src/plugins.ts [hunk 2]
```diff
@@ -70,4 +88,14 @@ export type FanoutHookFunction<
...
+
+export const tunnel = {
+  defineProvider: (input: {name: string}) => () => input,
+  startTunnel: (options: {provider: string; action: (port: number) => Promise<string | undefined>}) => {
+    return (inputs: {port: number; provider: string}) => {
+      if (inputs.provider !== options.provider) return undefined
+      return options.action(inputs.port)
+    }
+  },
```
Comment (isaacroldan) [inline packages/cli-kit/src/plugins.ts:100]: By having this middleware we can control which plugins are actually called.

Each plugin defines a provider and an action to be executed, then when the hook is called the action will be run only if the defined provider is the same as the intended provider (sent via runHook).

The types between this functions and `HookReturnsPerPlugin` are not the same, I'd need to find a way to tie them.

**Comment by isaacroldan** [comment]
@pepicrft @shauns After @pepicrft feedback on the ngrok plugin I added specific tunneling interfaces that makes it easier to create a tunnel plugin.

The changes are:
- Both hooks are now generic, `tunnel_start_{provider}` is no `tunnel_start`, all will be called, but only the ones that match the intended provider will be actually executed.
- Each plugin will need to implement these two hooks like this:

```ts
// provider.ts
export const TUNNEL_PROVIDER = 'ngrok'
export default plugins.tunnel.defineProvider({name: TUNNEL_PROVIDER})
```

```ts
// tunnel.ts
export default plugins.tunnel.startTunnel({ 
  provider: TUNNEL_PROVIDER, 
  action: async (port: number) => start({port})
})

async function start(.... // this handles the tunnel creating and returning the URL.
```

You can see examples in the updated [PR for the ngrok plugin](https://github.com/Shopify/awesome-shopify-cli-plugins/pull/31/files)


What do you think @shauns @pepicrft ? Do you like this approach more than the previous proposal? for reference:
```ts
const hook: plugins.FanoutHookFunction<'tunnel_provider', 'ngrok'> = async function () {
  return {hook: 'tunnel_start_ngrok', provider: 'ngrok'}
}
```

**Comment by pepicrft** [comment]
> What do you think @shauns @pepicrft ? Do you like this approach more than the previous proposal? for reference:

I like how idiomatic the new API is.

**Review by pepicrft — APPROVED** [review]

File: packages/cli-kit/src/plugins.ts [hunk 3]
```diff
@@ -5,12 +5,18 @@ import {PickByPrefix} from './typing/pick-by-prefix.js'
 import {MonorailEventPublic} from './monorail.js'
 import {Interfaces} from '@oclif/core'
 
+export type TunnelHook = `tunnel_start_${string}`
```
Comment (pepicrft) [inline packages/cli-kit/src/plugins.ts:8]: This `plugins.ts` is a good candidate to grow into a huge module. As soon as we see this one growing too much we should consider breaking it down into smaller and more manageable modules.

**Review by pepicrft — APPROVED** [review]

**Review by matteodepalo — APPROVED** [review]

**Pull request merged**